File: local/apps/android-overlay/src/ai/errandos/overlay/OverlayPresentation.java
package ai.errandos.overlay;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

final class OverlayPresentation {
    static final int VERSION = 1;

    static final class ProductChoice {
        final String offerId;
        final String title;
        final String spokenLabel;
        final String packSize;
        final String price;
        final String imageUrl;
        final String unitPrice;
        final String availabilityConstraint;
        final String recommendationLabel;

        ProductChoice(
            String offerId,
            String title,
            String spokenLabel,
            String packSize,
            String price
        ) {
            this(
                offerId,
                title,
                spokenLabel,
                packSize,
                price,
                null,
                null,
                null,
                null
            );
        }

        ProductChoice(
            String offerId,
            String title,
            String spokenLabel,
            String packSize,
            String price,
            String imageUrl,
            String unitPrice,
            String availabilityConstraint,
            String recommendationLabel
        ) {
            this.offerId = offerId;
            this.title = title;
            this.spokenLabel = spokenLabel;
            this.packSize = packSize;
            this.price = price;
            this.imageUrl = imageUrl;
            this.unitPrice = unitPrice;
            this.availabilityConstraint = availabilityConstraint;
            this.recommendationLabel = recommendationLabel;
        }
    }

    static final class CartLine {
        final String productId;
        final String name;
        final int quantity;
        final String unitPrice;
        final String lineTotal;

        CartLine(
            String productId,
            String name,
            int quantity,
            String unitPrice,
            String lineTotal
        ) {
            this.productId = productId;
            this.name = name;
            this.quantity = quantity;
            this.unitPrice = unitPrice;
            this.lineTotal = lineTotal;
        }
    }

    static final class CartSummary {
        final List<CartLine> lines;
        final String subtotal;
        final String addressLabel;
        final boolean verified;
        final boolean ordered;

        CartSummary(
            List<CartLine> lines,
            String subtotal,
            String addressLabel
        ) {
            this(lines, subtotal, addressLabel, false, true);
        }

        CartSummary(
            List<CartLine> lines,
            String subtotal,
            String addressLabel,
            boolean verified,
            boolean ordered
        ) {
            this.lines = Collections.unmodifiableList(
                new ArrayList<CartLine>(lines)
            );
            this.subtotal = subtotal;
            this.addressLabel = addressLabel;
            this.verified = verified;
            this.ordered = ordered;
        }

        boolean isVerifiedNotOrdered() {
            return verified && !ordered;
        }
    }

    static final class ProductSelectionBinding {
        final int version;
        final String clientId;
        final String taskId;
        final int taskRevision;
        final String interactionId;
        final String selectionId;
        final long expiresAtEpochMs;

        ProductSelectionBinding(
            int version,
            String clientId,
            String taskId,
            int taskRevision,
            String interactionId,
            String selectionId,
            long expiresAtEpochMs
        ) {
            this.version = version;
            this.clientId = clientId;
            this.taskId = taskId;
            this.taskRevision = taskRevision;
            this.interactionId = interactionId;
            this.selectionId = selectionId;
            this.expiresAtEpochMs = expiresAtEpochMs;
        }

        boolean isExpired(long nowEpochMs) {
            return nowEpochMs >= expiresAtEpochMs;
        }

        boolean sameSelection(ProductSelectionBinding other) {
            return other != null
                && selectionId.equals(other.selectionId)
                && interactionId.equals(other.interactionId)
                && taskId.equals(other.taskId)
                && taskRevision == other.taskRevision;
        }
    }

    static final class CompletionChoice {
        final String choiceId;
        final String label;
        final boolean enabled;
        final String disabledReason;

        CompletionChoice(
            String choiceId,
            String label,
            boolean enabled,
            String disabledReason
        ) {
            this.choiceId = choiceId;
            this.label = label;
            this.enabled = enabled;
            this.disabledReason = disabledReason;
        }
    }

    static final class CompletionInteraction {
        final int version;
        final String interactionId;
        final String taskId;
        final int taskRevision;
        final long expiresAtEpochMs;
        final String currentPaymentLabel;
        final List<CompletionChoice> choices;

        CompletionInteraction(
            int version,
            String interactionId,
            String taskId,
            int taskRevision,
            long expiresAtEpochMs,
            String currentPaymentLabel,
            List<CompletionChoice> choices
        ) {
            this.version = version;
            this.interactionId = interactionId;
            this.taskId = taskId;
            this.taskRevision = taskRevision;
            this.expiresAtEpochMs = expiresAtEpochMs;
            this.currentPaymentLabel = currentPaymentLabel;
            this.choices = Collections.unmodifiableList(
                new ArrayList<CompletionChoice>(choices)
            );
        }

        boolean isExpired(long nowEpochMs) {
            return nowEpochMs >= expiresAtEpochMs;
        }

        boolean sameInteraction(CompletionInteraction other) {
            return other != null
                && interactionId.equals(other.interactionId)
                && taskId.equals(other.taskId)
                && taskRevision == other.taskRevision;
        }
    }

    static final class TaskProgress {
        final int version;
        final String taskId;
        final String itemId;
        final String operationId;
        final String title;
        final String step;
        final String stage;
        final int sequence;
        final int currentItem;
        final int totalItems;
        final boolean cancellationAvailable;
        final String cancellationPolicy;
        final boolean terminal;

        TaskProgress(
            int version,
            String taskId,
            String itemId,
            String operationId,
            String title,
            String step,
            String stage,
            int sequence,
            int currentItem,
            int totalItems,
            boolean cancellationAvailable,
            String cancellationPolicy,
            boolean terminal
        ) {
            this.version = version;
            this.taskId = taskId;
            this.itemId = itemId;
            this.operationId = operationId;
            this.title = title;
            this.step = step;
            this.stage = stage;
            this.sequence = sequence;
            this.currentItem = currentItem;
            this.totalItems = totalItems;
            this.cancellationAvailable = cancellationAvailable;
            this.cancellationPolicy = cancellationPolicy;
            this.terminal = terminal;
        }

        String positionLabel() {
            if (currentItem < 1) return null;
            if (totalItems >= currentItem) {
                return currentItem + " of " + totalItems;
            }
            return "Item " + currentItem;
        }

        boolean hasKnownTotal() {
            return currentItem > 0 && totalItems >= currentItem;
        }

        String cancellationLabel() {
            if ("cancel_now".equals(cancellationPolicy)) {
                return "CANCEL AVAILABLE";
            }
            if ("stop_after_current_step".equals(cancellationPolicy)) {
                return "STOPS AFTER THIS STEP";
            }
            if ("reconcile_only".equals(cancellationPolicy)) {
                return "FINISHING VERIFICATION";
            }
            return terminal ? null : "CANNOT CANCEL";
        }
    }

    static final class Card {
        final String type;
        final String tone;
        final String headline;
        final String detail;
        final List<ProductChoice> options;
        final ProductSelectionBinding selection;
        final CompletionInteraction completionInteraction;
        final CartSummary cartSummary;
        final CompanionIssueV2 issue;

        Card(
            String type,
            String tone,
            String headline,
            String detail,
            List<ProductChoice> options,
            ProductSelectionBinding selection
        ) {
            this(
                type,
                tone,
                headline,
                detail,
                options,
                selection,
                null
            );
        }

        Card(
            String type,
            String tone,
            String headline,
            String detail,
            List<ProductChoice> options,
            ProductSelectionBinding selection,
            CartSummary cartSummary
        ) {
            this.type = type;
            this.tone = tone;
            this.headline = headline;
            this.detail = detail;
            this.options = Collections.unmodifiableList(
                new ArrayList<ProductChoice>(options)
            );
            this.selection = selection;
            this.completionInteraction = null;
            this.cartSummary = cartSummary;
            this.issue = null;
        }

        Card(
            String type,
            String tone,
            String headline,
            String detail,
            CompletionInteraction completionInteraction
        ) {
            this.type = type;
            this.tone = tone;
            this.headline = headline;
            this.detail = detail;
            this.options = Collections.emptyList();
            this.selection = null;
            this.completionInteraction = completionInteraction;
            this.cartSummary = null;
            this.issue = null;
        }

        Card(
            String type,
            String tone,
            String headline,
            String detail,
            CompanionIssueV2 issue
        ) {
            this.type = type;
            this.tone = tone;
            this.headline = headline;
            this.detail = detail;
            this.options = Collections.emptyList();
            this.selection = null;
            this.completionInteraction = null;
            this.cartSummary = null;
            this.issue = issue;
        }
    }

    final int version;
    final String mode;
    final String primarySurface;
    final String screenKind;
    final String screenRelevance;
    final String attentionSubject;
    final TaskProgress task;
    final Card card;
    final String spokenText;
    final String languageCode;
    final boolean autoCollapse;
    final long collapseAfterMs;
    final boolean keepVisibleWhileSpeaking;
    final boolean structured;

    OverlayPresentation(
        int version,
        String mode,
        String primarySurface,
        String screenKind,
        String screenRelevance,
        String attentionSubject,
        TaskProgress task,
        Card card,
        String spokenText,
        String languageCode,
        boolean autoCollapse,
        long collapseAfterMs,
        boolean keepVisibleWhileSpeaking,
        boolean structured
    ) {
        this.version = version;
        this.mode = mode;
        this.primarySurface = primarySurface;
        this.screenKind = screenKind;
        this.screenRelevance = screenRelevance;
        this.attentionSubject = attentionSubject;
        this.task = task;
        this.card = card;
        this.spokenText = spokenText;
        this.languageCode = languageCode;
        this.autoCollapse = autoCollapse;
        this.collapseAfterMs = collapseAfterMs;
        this.keepVisibleWhileSpeaking = keepVisibleWhileSpeaking;
        this.structured = structured;
    }

    static OverlayPresentation legacy(String reply, String state) {
        String safeReply = clean(reply, "Done.", 1000);
        String mode = modeFromLegacyState(state);
        Card card = new Card(
            "compact_status",
            toneFromMode(mode),
            headlineForMode(mode),
            safeReply,
            Collections.<ProductChoice>emptyList(),
            null
        );
        boolean persistent = "waiting_for_user".equals(mode)
            || "error".equals(mode)
            || "ambiguous".equals(mode);
        return new OverlayPresentation(
            VERSION,
            mode,
            "overlay_card",
            null,
            null,
            null,
            null,
            card,
            safeReply,
            "en-IN",
            !persistent,
            6500L,
            true,
            false
        );
    }

    static String clean(String value, String fallback, int maximumLength) {
        if (value == null) return fallback;
        String clean = value.trim();
        if (clean.isEmpty()) return fallback;
        return clean.length() <= maximumLength
            ? clean
            : clean.substring(0, maximumLength);
    }

    static String modeFromLegacyState(String state) {
        if ("listening".equals(state)) return "listening";
        if ("understanding".equals(state)) return "understanding";
        if ("verifying".equals(state)) return "verifying";
        if ("disconnected".equals(state)) return "disconnected";
        if ("paused".equals(state)) return "paused";
        if (
            "working".equals(state)
                || "searching".equals(state)
                || "adding".equals(state)
                || "checkout".equals(state)
        ) {
            return "acting";
        }
        if ("clarification".equals(state) || "confirmation".equals(state)) {
            return "waiting_for_user";
        }
        if ("success".equals(state)) return "success";
        if ("error".equals(state)) return "error";
        return "idle";
    }

    static String toneFromMode(String mode) {
        if ("success".equals(mode)) return "success";
        if ("error".equals(mode)) return "error";
        if ("ambiguous".equals(mode)) return "ambiguous";
        if (
            "disconnected".equals(mode)
                || "paused".equals(mode)
        ) {
            return "attention";
        }
        if ("waiting_for_user".equals(mode)) return "attention";
        if (
            "listening".equals(mode)
                || "understanding".equals(mode)
                || "reading".equals(mode)
                || "acting".equals(mode)
                || "verifying".equals(mode)
        ) {
            return "active";
        }
        return "neutral";
    }

    static String headlineForMode(String mode) {
        if ("listening".equals(mode)) return "LISTENING";
        if ("understanding".equals(mode)) return "UNDERSTANDING";
        if ("reading".equals(mode)) return "READING SCREEN";
        if ("acting".equals(mode)) return "WORKING";
        if ("verifying".equals(mode)) return "VERIFYING";
        if ("waiting_for_user".equals(mode)) return "YOUR TURN";
        if ("success".equals(mode)) return "DONE";
        if ("ambiguous".equals(mode)) return "CHECK NEEDED";
        if ("paused".equals(mode)) return "TASK PAUSED";
        if ("disconnected".equals(mode)) return "DISCONNECTED";
        if ("error".equals(mode)) return "NEEDS ATTENTION";
        return "JALDIAI";
    }

    boolean usesProviderScreen() {
        return "provider_screen".equals(primarySurface)
            && screenKind != null
            && screenRelevance != null
            && attentionSubject != null;
    }

    boolean needsContextCard() {
        if (task != null) return true;
        if (
            "product_choices".equals(card.type)
                || "cart_summary".equals(card.type)
                || "checkout_review".equals(card.type)
                || "changed_terms".equals(card.type)
                || "provider_constraint".equals(card.type)
                || "receipt".equals(card.type)
                || "ambiguous".equals(card.type)
                || "completion_choices".equals(card.type)
                || "companion_issue".equals(card.type)
        ) {
            return true;
        }
        return "waiting_for_user".equals(mode)
            && (
                "checkout_summary".equals(screenRelevance)
                    || "address_choices".equals(screenRelevance)
                    || "payment_selection".equals(screenRelevance)
            );
    }
}

File: local/packages/contracts/src/overlay-presentation.ts
import { z } from 'zod';
import {
  AndroidCartLineSchemaV1,
  AndroidCartReviewSchemaV1,
  AndroidCheckoutReviewSchemaV1,
  AndroidCurrentScreenSchemaV1,
  BlinkitProposalChangeSchemaV1,
} from './android-worker.js';
import { MoneySchema } from './proposals.js';

const OpaqueReferenceSchema = z.string().trim().min(1).max(200);
const LocalIdentifierSchema = (kind: string): z.ZodString => z.string()
  .regex(new RegExp(`^${kind}_[A-Za-z0-9-]{8,80}$`));

export const OverlayProductSelectionBindingSchemaV1 = z.object({
  version: z.literal(1),
  clientId: OpaqueReferenceSchema,
  taskId: LocalIdentifierSchema('task'),
  taskRevision: z.number().int().nonnegative(),
  clarificationId: LocalIdentifierSchema('clarification'),
  selectionId: LocalIdentifierSchema('selection'),
  expiresAt: z.string().datetime(),
}).strict();

export const OverlayProductSelectionBindingSchemaV2 = z.object({
  version: z.literal(2),
  clientId: OpaqueReferenceSchema,
  taskId: LocalIdentifierSchema('task'),
  taskRevision: z.number().int().nonnegative(),
  interactionId: OpaqueReferenceSchema,
  selectionId: LocalIdentifierSchema('selection'),
  expiresAt: z.string().datetime(),
}).strict();

export const OverlayProductSelectionBindingSchema =
  z.discriminatedUnion('version', [
    OverlayProductSelectionBindingSchemaV1,
    OverlayProductSelectionBindingSchemaV2,
  ]);

export const OverlayModeSchemaV1 = z.enum([
  'idle',
  'listening',
  'understanding',
  'reading',
  'acting',
  'verifying',
  'waiting_for_user',
  'success',
  'error',
  'ambiguous',
]);

export const OverlayPrimarySurfaceSchemaV1 = z.enum([
  'provider_screen',
  'overlay_card',
]);

export const OverlayScreenRelevanceSchemaV1 = z.enum([
  'product_options',
  'product_detail',
  'cart_summary',
  'checkout_summary',
  'payment_selection',
  'address_choices',
  'order_confirmation',
  'order_history',
  'authentication',
]);

export const OverlayAttentionSubjectSchemaV1 = z.enum([
  'options',
  'product',
  'cart',
  'checkout',
  'payment',
  'address',
  'confirmation',
  'recent_orders',
  'authentication',
]);

export const OverlayToneSchemaV1 = z.enum([
  'neutral',
  'active',
  'attention',
  'success',
  'error',
  'ambiguous',
  'confirmation',
]);

const OverlayProductImageUrlSchemaV1 = z.string().trim().url().max(2_048)
  .refine((value) => {
    try {
      const url = new URL(value);
      const hostname = url.hostname.toLowerCase();
      return url.protocol === 'https:'
        && (hostname === 'blinkit.com'
        || hostname.endsWith('.blinkit.com')
        || hostname === 'grofers.com'
        || hostname.endsWith('.grofers.com'));
    } catch {
      return false;
    }
  }, 'Product image URL must use HTTPS on a Blinkit-owned domain');

const OverlayUnitPriceSchemaV1 = z.object({
  ...MoneySchema.shape,
  unit: z.string().trim().min(1).max(40),
}).strict();

const OverlayProductRecommendationSchemaV1 = z.object({
  code: z.string().regex(/^[a-z][a-z0-9_]{1,63}$/),
  label: z.string().trim().min(1).max(100),
}).strict();

export const OverlayProductChoiceSchemaV1 = z.object({
  offerId: OpaqueReferenceSchema,
  title: z.string().trim().min(1).max(300),
  spokenLabel: z.string().trim().min(1).max(300),
  packSize: z.string().trim().min(1).max(100).optional(),
  price: MoneySchema.optional(),
  imageUrl: OverlayProductImageUrlSchemaV1.optional(),
  unitPrice: OverlayUnitPriceSchemaV1.optional(),
  availabilityConstraint: z.string().trim().min(1).max(160).optional(),
  recommendation: OverlayProductRecommendationSchemaV1.optional(),
}).strict();

const OverlayVerifiedCartLineSchemaV1 = z.object({
  ...AndroidCartLineSchemaV1.shape,
  spokenLabel: z.string().trim().min(1).max(300).optional(),
  packSize: z.string().trim().min(1).max(100).optional(),
}).strict();

export const OverlayVerifiedCartSummarySchemaV1 = z.object({
  verified: z.literal(true),
  lines: z.array(OverlayVerifiedCartLineSchemaV1).min(1).max(30),
  subtotal: MoneySchema,
  addressLabel: AndroidCartReviewSchemaV1.shape.addressLabel,
}).strict().superRefine((cart, context) => {
  const minor = (amount: number): number => Math.round(amount * 100);
  let expectedSubtotal = 0;
  cart.lines.forEach((line, index) => {
    const expectedLineTotal = minor(line.unitPrice.amount) * line.quantity;
    const actualLineTotal = minor(line.lineTotal.amount);
    if (actualLineTotal !== expectedLineTotal) {
      context.addIssue({
        code: 'custom',
        message: 'lineTotal must equal unitPrice times quantity',
        path: ['lines', index, 'lineTotal'],
      });
    }
    expectedSubtotal += actualLineTotal;
  });
  if (minor(cart.subtotal.amount) !== expectedSubtotal) {
    context.addIssue({
      code: 'custom',
      message: 'subtotal must equal the sum of line totals',
      path: ['subtotal'],
    });
  }
});

export const OverlayCardSchemaV1 = z.discriminatedUnion('type', [
  z.object({
    type: z.literal('compact_status'),
    tone: OverlayToneSchemaV1,
  }).strict(),
  z.object({
    type: z.literal('product_choices'),
    options: z.array(OverlayProductChoiceSchemaV1).min(1).max(10),
    selection: OverlayProductSelectionBindingSchema.optional(),
  }).strict(),
  z.object({
    type: z.literal('cart_summary'),
    ordered: z.literal(false),
    cart: OverlayVerifiedCartSummarySchemaV1,
  }).strict(),
  z.object({
    type: z.literal('checkout_review'),
    checkout: AndroidCheckoutReviewSchemaV1,
    ordered: z.literal(false),
  }).strict(),
  z.object({
    type: z.literal('changed_terms'),
    changes: z.array(BlinkitProposalChangeSchemaV1).min(1).max(8),
  }).strict(),
  z.object({
    type: z.literal('provider_constraint'),
    reason: z.string().trim().min(1).max(300),
  }).strict(),
  z.object({
    type: z.literal('receipt'),
    providerReference: OpaqueReferenceSchema,
  }).strict(),
  z.object({
    type: z.literal('ambiguous'),
    reconciliationId: OpaqueReferenceSchema.optional(),
  }).strict(),
]);

const VerifiedCurrentScreenSchemaV1 = z.object({
  ...AndroidCurrentScreenSchemaV1.shape,
  relevance: OverlayScreenRelevanceSchemaV1,
  verified: z.literal(true),
}).strict();

const AttentionCueSchemaV1 = z.object({
  instruction: z.literal('check_current_screen'),
  subject: OverlayAttentionSubjectSchemaV1,
}).strict();

export const OverlayTaskProgressStageSchemaV1 = z.enum([
  'queued',
  'waiting_for_provider',
  'searching',
  'waiting_for_choice',
  'adding',
  'verifying',
  'reconciling',
  'completed',
  'failed',
  'cancelled',
  'ambiguous',
]);

export const OverlayTaskCancellationPolicySchemaV1 = z.enum([
  'cancel_now',
  'stop_after_current_step',
  'reconcile_only',
  'not_cancellable',
]);

const OverlayStructuredTaskProgressSchemaV1 = z.object({
  version: z.literal(1),
  taskId: LocalIdentifierSchema('task'),
  itemId: LocalIdentifierSchema('task_item').optional(),
  operationId: LocalIdentifierSchema('operation'),
  title: z.string().trim().min(1).max(120),
  step: z.string().trim().min(1).max(200),
  stage: OverlayTaskProgressStageSchemaV1,
  sequence: z.number().int().nonnegative(),
  position: z.object({
    current: z.number().int().positive(),
    total: z.number().int().positive().optional(),
  }).strict().optional(),
  cancellation: z.object({
    available: z.boolean(),
    policy: OverlayTaskCancellationPolicySchemaV1,
  }).strict(),
  terminal: z.boolean(),
}).strict().superRefine((value, context) => {
  const terminalStage = [
    'completed',
    'failed',
    'cancelled',
    'ambiguous',
  ].includes(value.stage);
  if (value.terminal !== terminalStage) {
    context.addIssue({
      code: 'custom',
      message: 'terminal must match the progress stage',
      path: ['terminal'],
    });
  }
  if (
    value.position?.total !== undefined
    && value.position.current > value.position.total
  ) {
    context.addIssue({
      code: 'custom',
      message: 'current position cannot exceed total',
      path: ['position', 'current'],
    });
  }
  if (
    value.cancellation.available
      !== (
        value.cancellation.policy === 'cancel_now'
        || value.cancellation.policy === 'stop_after_current_step'
      )
  ) {
    context.addIssue({
      code: 'custom',
      message: 'cancellation availability must match its policy',
      path: ['cancellation', 'available'],
    });
  }
});

// Accepted for wire compatibility with presentation-v1 producers that shipped
// before semantic execution checkpoints. New producers must emit the versioned
// shape above; native rendering treats this legacy shape as text-only.
const OverlayLegacyTaskProgressSchemaV1 = z.object({
  title: z.string().trim().min(1).max(120),
  step: z.string().trim().min(1).max(200).optional(),
  progress: z.number().min(0).max(1).optional(),
}).strict();

export const OverlayPresentationSchemaV1 = z.object({
  version: z.literal(1),
  mode: OverlayModeSchemaV1,
  task: z.union([
    OverlayStructuredTaskProgressSchemaV1,
    OverlayLegacyTaskProgressSchemaV1,
  ]).optional(),
  primarySurface: OverlayPrimarySurfaceSchemaV1,
  currentScreen: VerifiedCurrentScreenSchemaV1.optional(),
  attentionCue: AttentionCueSchemaV1.optional(),
  card: OverlayCardSchemaV1,
  spoken: z.object({
    text: z.string().trim().min(1).max(1_000),
    languageCode: z.string().regex(/^[a-z]{2,3}-[A-Z]{2}$/),
  }).strict(),
  behavior: z.object({
    autoCollapse: z.boolean(),
    collapseAfterMs: z.number().int().min(0).max(60_000).optional(),
    keepVisibleWhileSpeaking: z.boolean(),
  }).strict(),
}).strict().superRefine((value, context) => {
  if (value.currentScreen?.kind === 'unknown') {
    context.addIssue({
      code: 'custom',
      message: 'A verified current screen must have a recognized kind',
      path: ['currentScreen', 'kind'],
    });
  }

  if (value.primarySurface === 'provider_screen') {
    if (!value.currentScreen) {
      context.addIssue({
        code: 'custom',
        message: 'Provider-screen presentation requires a verified current screen',
        path: ['currentScreen'],
      });
    }
    if (!value.attentionCue) {
      context.addIssue({
        code: 'custom',
        message: 'Provider-screen presentation requires an attention cue',
        path: ['attentionCue'],
      });
    }
  } else if (value.attentionCue) {
    context.addIssue({
      code: 'custom',
      message: 'Overlay-card presentation cannot direct attention to the provider screen',
      path: ['attentionCue'],
    });
  }

  if (value.mode === 'ambiguous' && value.card.type !== 'ambiguous') {
    context.addIssue({
      code: 'custom',
      message: 'Ambiguous mode requires an ambiguous card',
      path: ['card', 'type'],
    });
  }
});

export type OverlayModeV1 = z.infer<typeof OverlayModeSchemaV1>;
export type OverlayPrimarySurfaceV1 = z.infer<typeof OverlayPrimarySurfaceSchemaV1>;
export type OverlayScreenRelevanceV1 = z.infer<typeof OverlayScreenRelevanceSchemaV1>;
export type OverlayAttentionSubjectV1 = z.infer<typeof OverlayAttentionSubjectSchemaV1>;
export type OverlayToneV1 = z.infer<typeof OverlayToneSchemaV1>;
export type OverlayProductChoiceV1 = z.infer<typeof OverlayProductChoiceSchemaV1>;
export type OverlayVerifiedCartSummaryV1 = z.infer<
  typeof OverlayVerifiedCartSummarySchemaV1
>;
export type OverlayProductSelectionBindingV1 = z.infer<
  typeof OverlayProductSelectionBindingSchemaV1
>;
export type OverlayProductSelectionBindingV2 = z.infer<
  typeof OverlayProductSelectionBindingSchemaV2
>;
export type OverlayProductSelectionBinding = z.infer<
  typeof OverlayProductSelectionBindingSchema
>;
export type OverlayTaskProgressStageV1 = z.infer<
  typeof OverlayTaskProgressStageSchemaV1
>;
export type OverlayTaskCancellationPolicyV1 = z.infer<
  typeof OverlayTaskCancellationPolicySchemaV1
>;
export type OverlayStructuredTaskProgressV1 = z.infer<
  typeof OverlayStructuredTaskProgressSchemaV1
>;
export type OverlayCardV1 = z.infer<typeof OverlayCardSchemaV1>;
export type OverlayPresentationV1 = z.infer<typeof OverlayPresentationSchemaV1>;
